[SYCL] Fix ICE when -g and -ffile-prefix-map flags are set

zahiraam opened the pull request:
Set the source file name when footer is present and option fmacro-prefix-map is used when generating debug information.

This patch is to fix a crash in the compiler when generating debug information with the option -ffile-prefix-map.
Example:
$ pwd
/tmp/tmp.jocQk9WVa0
$
$ cat builtin.cpp 
void a(__builtin_va_list);
using ::a;
$
$ clang -c -g -fsycl /tmp/tmp.jocQk9WVa0/builtin.cpp
$ 
$ clang -g -fsycl -ffile-prefix-map=/tmp/= /tmp/tmp.jocQk9WVa0/builtin.cpp 
Expected<T> must be checked before access or destruction.
Unchecked Expected<T> contained error:
No such file or directoryPLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: 
…………………/tmp/tmp.jocQk9WVa0/builtin.cpp
1.	<eof> parser at end of file
2.	/tmp/tmp.jocQk9WVa0/builtin.cpp:2:9: LLVM IR generation of declaration 'a

https://github.com/intel/llvm/blob/sycl/clang/lib/CodeGen/CGDebugInfo.cpp#L397 is returning a error because the FileName is wrong.

elizabethandrews commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -411,6 +411,14 @@ llvm::DIFile *CGDebugInfo::getOrCreateFile(SourceLocation Loc) {
+      // When fmacro-prefix-map is used, the original source file
+      // file name is indicated by FullMainFileName instead of the CU.
+      auto &CGO = CGM.getCodeGenOpts();
+      FileName = CGO.FullMainFileName;
```
What is FullMainFileName?

elizabethandrews commented on clang/test/CodeGenSYCL/debug-info-file-prefix-map.cpp:
```diff
@@ -0,0 +1,14 @@
+// RUN: -fmacro-prefix-map=%S/Inputs/= -fcoverage-prefix-map=%S/Inputs/= \
+// RUN: -fdebug-prefix-map=%S/Inputs/= \
+// RUN: -debug-info-kind=constructor -emit-llvm -O0 -o - "%S/Inputs/builtin.cpp" \
+// RUN: | FileCheck %s
```
Please add a description for this test

zahiraam replied:
FullMainFileName is declared here [llvm/ llvm/CGDebugInfo.cpp at sycl · intel/llvm · GitHubCGDebugInfo.cpp at sycl · intel/llvm · GitHub](https://github.com/intel/llvm/blob/sycl/clang/lib/CodeGen/CGDebugInfo.cpp#L551) and represents the name of the original source file name.

zahiraam replied:
It's a codegen option triggered by https://github.com/intel/llvm/blob/sycl/clang/include/clang/Driver/Options.td#L6333

elizabethandrews replied:
Sorry if I wasn't clear earlier. I meant a description in this file. All test files need a comment describing what the test does.

elizabethandrews reviewed (commented):
Can you describe what this change is doing in more detail in PR description?

premanandrao commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -411,6 +411,14 @@ llvm::DIFile *CGDebugInfo::getOrCreateFile(SourceLocation Loc) {
     // createFile() below for canonicalization if the source file was specified
     // with an absolute path.
     FileName = TheCU->getFile()->getFilename();
```
Can you move the above assignment to an else clause?

premanandrao commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -411,6 +411,14 @@ llvm::DIFile *CGDebugInfo::getOrCreateFile(SourceLocation Loc) {
+    }
   } else {
     PresumedLoc PLoc = SM.getPresumedLoc(Loc);
     FileName = PLoc.getFilename();
```
Line 429 below is not needed.

elizabethandrews commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -407,18 +407,25 @@ llvm::DIFile *CGDebugInfo::getOrCreateFile(SourceLocation Loc) {
 
     if (FileName.empty()) {
       FileName = TheCU->getFile()->getFilename();
-    } else {
```
Why was this removed?

zahiraam replied:
This was a simplification that Prem had proposed in the comments above.

elizabethandrews replied:
I'm not sure if I am missing something. What is FileName set to if `!FileName.empty()`

zahiraam replied:
FileName is defined at line 425 above. If it's empty we computed it from the TheCU, otherwise we keep going.

premanandrao replied:
@elizabethandrews, the else clause was setting it to the same value that the assignment on line 425 was doing, so it was redundant.

zahiraam commented:
@bwyma Thanks for reviewing and checking the debug information. I have fixed the issue you raised about the directory path. Also added in the other LIT tests a check line for the directory to make sure it's correct (instead of using regular expression). Would you please check if this new patch fixes the issue? Thanks.

elizabethandrews commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -656,9 +663,13 @@ void CGDebugInfo::CreateCompileUnit() {
-      remapDIPath(MainFileName), remapDIPath(getCurrentDirname()), CSInfo,
-      getSource(SM, SM.getMainFileID()));
+
+  if (!CGM.getCodeGenOpts().SYCLUseMainFileName)
```
Please add comments describing your change. Is the comment on L660 still accurate?

zahiraam commented:
> LGTM.

Thanks.

bader commented on clang/test/CodeGenSYCL/Inputs/builtin.cpp:
```diff
@@ -0,0 +1,3 @@
+void a(__builtin_va_list);
+using ::a;
+
```
I believe we can remove this file and move the content to `clang/test/CodeGenSYCL/debug-info-file-prefix-map.cpp` directly.

zahiraam commented:
@intel/dpcpp-devops-reviewers I have been re-basing and re-testing this PR for a week. Can someone help with the latest fail please? Thanks.
It says:
"This request was automatically failed because there were no enabled runners online to process the request for more than 1 days."

aelovikov-intel commented:
@apstasen , could you take a look please? We've seen similar issue with not getting AWS CUDA runner picked up in at least one another PR.

bader commented:
> @intel/dpcpp-devops-reviewers I have been re-basing and re-testing this PR for a week. Can someone help with the latest fail please? Thanks. It says: "This request was automatically failed because there were no enabled runners online to process the request for more than 1 days."

I hope https://github.com/intel/llvm/pull/8567 fixes this issue.

zahiraam commented:
> > @intel/dpcpp-devops-reviewers I have been re-basing and re-testing this PR for a week. Can someone help with the latest fail please? Thanks. It says: "This request was automatically failed because there were no enabled runners online to process the request for more than 1 days."
> 
> I hope #8567 fixes this issue.

The testing started  14h 14m 8s ago. Unlless I am mistaken this is still hanging?
"Requested labels: aws-cuda_4318684912-4
Job defined at: intel/llvm/.github/workflows/sycl_linux_build_and_test.yml@refs/heads/sycl
Reusable workflow chain:
intel/llvm/.github/workflows/sycl_precommit.yml@refs/heads/sycl (fe12caba47874ac522953bfec56f740e1a823996)
-> intel/llvm/.github/workflows/sycl_linux_build_and_test.yml@refs/heads/sycl (fe12caba47874ac522953bfec56f740e1a823996)
Waiting for a runner to pick up this job..."
I seem to have no way of re-starting the test again. 
@bader, @Stasenko, Alexander P?

bader commented:
@zahiraam, you need to merge the fix to your `DBI` branch.

zahiraam commented:
@bader Your fix seems to have worked. Thanks.
Can we now merge in this PR?
